go/pkg/pwagent/daemon.go

```diff
@@ -56,28 +64,121 @@ func Daemon(ctx context.Context, cli *client.Client, logger *zap.Logger) error {
+	for {
+		// fetch local info from docker daemon
+		pathwarInfo, err := pwcompose.GetPathwarInfo(ctx, cli)
+		if err != nil {
+			return errcode.ErrComposeGetPathwarInfo.Wrap(err)
+		}
+		fmt.Println("info", godev.PrettyJSON(pathwarInfo))
 
-	// compute the difference
+		agentOpts := AgentOpts{
+			DomainSuffix:      "local",
+			HostIP:            "0.0.0.0",
+			HostPort:          "8000",
+			ModeratorPassword: "",
+			Salt:              "",
+			AllowedUsers:      map[string][]int64{},
+			ForceRecreate:     false,
+			NginxDockerImage:  "docker.io/library/nginx:stable-alpine",
+		}
+		// compute the difference
+		for _, apiInstance := range apiInstances.GetInstances() {
+			found := false
+			needRedump := false
+			for _, flavor := range pathwarInfo.RunningFlavors {
+				apiInstanceFlavor := apiInstance.GetFlavor()
+				if apiInstanceFlavor != nil {
+					apiInstanceFlavorChallenge := apiInstanceFlavor.GetChallenge()
+					if apiInstanceFlavorChallenge != nil {
+						if flavor.Name == apiInstanceFlavorChallenge.GetName() && flavor.Version == apiInstanceFlavor.GetVersion() {
+							found = true
+							if apiInstance.GetStatus() == pwdb.ChallengeInstance_NeedRedump {
+								needRedump = true
+							}
+						}
+					}
+					seasonChallenges := apiInstanceFlavor.GetSeasonChallenges()
+					if seasonChallenges != nil {
+						for _, seasonChallenge := range seasonChallenges {
+							subscriptions := seasonChallenge.GetSubscriptions()
+							if subscriptions != nil {
+								for _, subscription := range subscriptions {
+									if subscription.GetStatus() == pwdb.ChallengeSubscription_Active {
+										team := subscription.GetTeam()
+										if team != nil {
+											members := team.GetMembers()
+											if members != nil {
+												for _, member := range members {
+													for _, instance := range flavor.Instances {
+														for _, port := range instance.Ports {
+															if port.PublicPort != 0 {
+																// configure nginx
+																// - compute the list of members in Active teams
+																// - generate a hash based on their userIDs
+																instanceName := instance.Names[0][1:]
+																_, entryFound := agentOpts.AllowedUsers[instanceName]
+																if !entryFound {
+																	agentOpts.AllowedUsers[instanceName] = []int64{member.GetID()}
+																} else {
+																	allowedUsersSlice := agentOpts.AllowedUsers[instanceName]
+																	allowedUsersSlice = append(allowedUsersSlice, member.GetID())
+																	agentOpts.AllowedUsers[instanceName] = allowedUsersSlice
+																}
+															}
+														}
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
 
-	// start missing instances
+			// start missing instances
+			// redump instances with state=NeedRedump
+			if !found || needRedump {
+				composeFile, err := pwcompose.Prepare(
```

Comment:
Ok, I wasn't clear:

* prepare is made once by a challenge developer, then he/she provides us the output string of prepare so we can add it in our API
* agent NEVER builds a challenge itself, it only starts "ready-to-use" challenges

Reply:
For now, we miss a dedicated field on ChallengeFlavor to put the result of `compose prepare`, please use this one temporarily in the temporary fake data at the top of this function: 

https://github.com/pathwar/pathwar/blob/master/api/pwdb.proto#L63

```go
ChallengeFlavor{
  Version: "latest",
  SourceURL: "result of compose prepare",
}
```

Reply:
sourceURL looks more like an url to retrieve the data..
```
				resp, err := http.Get(apiInstance.GetFlavor().GetSourceURL())
				if err != nil {
					return errcode.TODO.Wrap(err)
				}
				defer resp.Body.Close()
				preparedComposeData, err := ioutil.ReadAll(resp.Body)

				err = pwcompose.Up(ctx, string(preparedComposeData), "default", needRedump, &configData, cli, logger)
				if err != nil {
					return errcode.TODO.Wrap(err)
				}
```